### DISCUSSION TITLE
[Feature Request]: Populate requestAsyncStorage with more data

### DISCUSSION THREAD

**Discussion by jeengbe** [vercel/next.js#47976] [Ideas]
### Goals

To empower the user and allow for more flexibility, more data should be available the same way `headers()` and `cookies()` is. 

I know only of #47108, but I assume there are countless more issues and users who would want this.

Specifically, I am thinking about the following properties:
- `NextRequest.nextUrl`
- `NextRequest.geo`
- TBD

### Non-Goals

n/a

### Background

Without further research, I have very quickly discovered the aforementioned issue (#47108), which also came with [a work-around](https://github.com/vercel/next.js/issues/47108#issuecomment-1480431453) for this very specific problem. While most properties from the request could be serialized and put into a header, this is a less than optimal solution. 

The only reason I could think of this not being implemented yet is a potentially too high overhead? I am not familiar with how async storage works precisely.

### Proposal

I have traced the context's creation back to:
https://github.com/vercel/next.js/blob/a7837bd09046719e0f6fbc43a9c3da4c5e0ede69/packages/next/src/server/async-storage/request-async-storage-wrapper.ts#L67 

A brief look at the two places where this method is called shows that the required information is available. https://github.com/vercel/next.js/blob/a7837bd09046719e0f6fbc43a9c3da4c5e0ede69/packages/next/src/server/future/route-modules/app-route/module.ts#L264 https://github.com/vercel/next.js/blob/a7837bd09046719e0f6fbc43a9c3da4c5e0ede69/packages/next/src/server/app-render/app-render.tsx#L1539

I would be interested in working on this myself.

**Comment by michal-izewski** [comment]
That is exactly what I'm looking for, I'm struggling with passing user context from middleware to the page components (when using App Router), and passing it serialized in headers is something I'm trying to avoid
